common/trinary/trit_tryte_sse42.h

```diff
@@ -0,0 +1,281 @@
+                                  _mm_setr_epi8(REPEAT(10, setMSB) COMMA(1) INDEX_3DIFF_0F COMMA(0) REPEAT(0, setMSB))};
+  /* The mask with interleaved bytes of 0xFF and 0x00 */
+  const __m128i byteInterMask = _mm_set_epi32(0xFF00FF00, 0xFF00FF00, 0xFF00FF00, 0xFF00FF00);
+
+  /* Start converting */
+  for (size_t i = 0; i < length / 3 / block_8bit; i++) {
+    /* Get trit data */
+    __m128i dataFirst = _mm_loadu_si128((__m128i *)(trits) + i * 3);
+    __m128i dataMid = _mm_loadu_si128((__m128i *)(trits) + i * 3 + 1);
+    __m128i dataLast = _mm_loadu_si128((__m128i *)(trits) + i * 3 + 2);
+    /*
+     * Each block represents a trit.
+     *                                           shuffle
+     *             ----------------        ------                  ------     ------     ------
+     * dataFirst = | a1 | a2 | a3 | ...... | f1 |       lowTrit  = | a1 | ... | f1 | ... | p1 |
+     *             ----------------        ------                  ------     ------     ------
+     *             ----------------        ------                  ------     ------     ------
+     * dataMid   = | f2 | f3 | g1 | ...... | k2 |  =>   midTrit  = | a2 | ... | f2 | ... | p2 |
+     *             ----------------        ------                  ------     ------     ------
+     *             ----------------        ------                  ------     ------     ------
+     * dataLast  = | k3 | l1 | l2 | ...... | p3 |       highTrit = | a3 | ... | f3 | ... | p3 |
+     *             ----------------        ------                  ------     ------     ------
+     */
+    __m128i lowTrit =
+        _mm_or_si128(_mm_shuffle_epi8(dataFirst, shuffleLow[0]),
+                     _mm_or_si128(_mm_shuffle_epi8(dataMid, shuffleLow[1]), _mm_shuffle_epi8(dataLast, shuffleLow[2])));
+    __m128i midTrit =
+        _mm_or_si128(_mm_shuffle_epi8(dataFirst, shuffleMid[0]),
+                     _mm_or_si128(_mm_shuffle_epi8(dataMid, shuffleMid[1]), _mm_shuffle_epi8(dataLast, shuffleMid[2])));
+    __m128i highTrit = _mm_or_si128(
+        _mm_shuffle_epi8(dataFirst, shuffleHigh[0]),
+        _mm_or_si128(_mm_shuffle_epi8(dataMid, shuffleHigh[1]), _mm_shuffle_epi8(dataLast, shuffleHigh[2])));
+    /* lowResult = (lowTrit) */
+    __m128i lowResult = lowTrit;
+    /* midResult = (midTrit * 3) */
+    __m128i midResult = _mm_or_si128(
+        _mm_and_si128(byteInterMask,
+                      _mm_mullo_epi16(_mm_and_si128(midTrit, byteInterMask),
+                                      _mm_set_epi16(0x0003, 0x0003, 0x0003, 0x0003, 0x0003, 0x0003, 0x0003, 0x0003))),
+        _mm_andnot_si128(byteInterMask, _mm_mullo_epi16(_mm_and_si128(midTrit, ~byteInterMask),
+                                                        _mm_set_epi16(0x0003, 0x0003, 0x0003, 0x0003, 0x0003, 0x0003,
+                                                                      0x0003, 0x0003))));
+    /* highResult = (highTrit * 9) */
+    __m128i highResult = _mm_or_si128(
```

Comment:
Why `*9` and `*3` operations are so comlex? `x*3` is just `x+x+x` - just two _mm_add_epi8 instructions, and `x*9=x+2*2*2*x` - four add instructions.

Reply:
I'll rewrite with `_mm_add_epi8` intrinsic function.

Reply:
Done.